File: src/main/java/com/pixelreel/service/ExternalSearchService.java
package com.pixelreel.service;

import com.pixelreel.dto.external.ExternalGameSearchResult;
import com.pixelreel.dto.external.ExternalMovieSearchResult;
import com.pixelreel.dto.external.ExternalSearchResponse;
import com.pixelreel.dto.external.ProviderSearchResult;
import com.pixelreel.service.search.GameSearchProvider;
import com.pixelreel.service.search.MovieSearchProvider;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Arrays;
import java.util.Set;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

@Service
@RequiredArgsConstructor
public class ExternalSearchService {
  private final List<MovieSearchProvider> movieSearchProviders;
  private final List<GameSearchProvider> gameSearchProviders;

  public ExternalSearchResponse<ExternalMovieSearchResult> searchMovies(
      String query,
      int page,
      List<String> providers
  ) {
    ExternalSearchResponse<ExternalMovieSearchResult> response = new ExternalSearchResponse<>();
    response.setQuery(query);
    response.setPage(page);
    response.setProviders(runMovieProviders(query, page, providers));
    return response;
  }

  public ExternalSearchResponse<ExternalGameSearchResult> searchGames(
      String query,
      int page,
      List<String> providers
  ) {
    ExternalSearchResponse<ExternalGameSearchResult> response = new ExternalSearchResponse<>();
    response.setQuery(query);
    response.setPage(page);
    response.setProviders(runGameProviders(query, page, providers));
    return response;
  }

  private List<ProviderSearchResult<ExternalMovieSearchResult>> runMovieProviders(
      String query,
      int page,
      List<String> providers
  ) {
    List<MovieSearchProvider> selected = selectMovieProviders(providers);
    List<ProviderSearchResult<ExternalMovieSearchResult>> results = new ArrayList<>();
    for (MovieSearchProvider provider : selected) {
      results.add(safeMovieSearch(provider, query, page));
    }
    return results;
  }

  private List<ProviderSearchResult<ExternalGameSearchResult>> runGameProviders(
      String query,
      int page,
      List<String> providers
  ) {
    List<GameSearchProvider> selected = selectGameProviders(providers);
    List<ProviderSearchResult<ExternalGameSearchResult>> results = new ArrayList<>();
    for (GameSearchProvider provider : selected) {
      results.add(safeGameSearch(provider, query, page));
    }
    return results;
  }

  private List<MovieSearchProvider> selectMovieProviders(List<String> providers) {
    if (providers == null || providers.isEmpty()) {
      return movieSearchProviders;
    }
    Set<String> set = providers.stream()
        .filter(StringUtils::hasText)
        .flatMap(value -> Arrays.stream(value.split(",")))
        .map(value -> value.trim().toLowerCase(Locale.ROOT))
        .filter(StringUtils::hasText)
        .collect(Collectors.toSet());
    return movieSearchProviders.stream()
        .filter(provider -> set.contains(provider.id().toLowerCase(Locale.ROOT)))
        .toList();
  }

  private List<GameSearchProvider> selectGameProviders(List<String> providers) {
    if (providers == null || providers.isEmpty()) {
      return gameSearchProviders;
    }
    Set<String> set = providers.stream()
        .filter(StringUtils::hasText)
        .flatMap(value -> Arrays.stream(value.split(",")))
        .map(value -> value.trim().toLowerCase(Locale.ROOT))
        .filter(StringUtils::hasText)
        .collect(Collectors.toSet());
    return gameSearchProviders.stream()
        .filter(provider -> set.contains(provider.id().toLowerCase(Locale.ROOT)))
        .toList();
  }

  private ProviderSearchResult<ExternalMovieSearchResult> safeMovieSearch(
      MovieSearchProvider provider,
      String query,
      int page
  ) {
    try {
      return provider.search(query, page);
    } catch (Exception ex) {
      ProviderSearchResult<ExternalMovieSearchResult> result = new ProviderSearchResult<>();
      result.setProvider(provider.id());
      result.setEnabled(true);
      result.setMessage("搜索失败: " + ex.getMessage());
      result.setPage(page);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }
  }

  private ProviderSearchResult<ExternalGameSearchResult> safeGameSearch(
      GameSearchProvider provider,
      String query,
      int page
  ) {
    try {
      return provider.search(query, page);
    } catch (Exception ex) {
      ProviderSearchResult<ExternalGameSearchResult> result = new ProviderSearchResult<>();
      result.setProvider(provider.id());
      result.setEnabled(true);
      result.setMessage("搜索失败: " + ex.getMessage());
      result.setPage(page);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }
  }
}

File: express-backend/src/services/ExternalSearchService.ts
import {
  ExternalMovieSearchResult,
  ExternalGameSearchResult,
  ExternalTvShowSearchResult,
  ExternalSearchResponse,
  ProviderSearchResult,
} from '../dto/external-search';
import { MovieSearchProvider } from './provider/MovieSearchProvider';
import { GameSearchProvider } from './provider/GameSearchProvider';
import { TvShowSearchProvider } from './provider/TvShowSearchProvider';

import { TmdbMovieSearchProvider } from './provider/TmdbMovieSearchProvider';
import { OmdbMovieSearchProvider } from './provider/OmdbMovieSearchProvider';
import { TraktMovieSearchProvider } from './provider/TraktMovieSearchProvider';
import { DoubanMovieSearchProvider } from './provider/DoubanMovieSearchProvider';
import { ImdbMovieSearchProvider } from './provider/ImdbMovieSearchProvider';

import { RawgGameSearchProvider } from './provider/RawgGameSearchProvider';
import { SteamGameSearchProvider } from './provider/SteamGameSearchProvider';
import { PsnGameSearchProvider } from './provider/PsnGameSearchProvider';
import { XboxGameSearchProvider } from './provider/XboxGameSearchProvider';
import { SwitchGameSearchProvider } from './provider/SwitchGameSearchProvider';

import { TmdbTvShowSearchProvider } from './provider/TmdbTvShowSearchProvider';

// 外部搜索聚合服务

const movieProviders: MovieSearchProvider[] = [
  new TmdbMovieSearchProvider(),
  new OmdbMovieSearchProvider(),
  new TraktMovieSearchProvider(),
  new DoubanMovieSearchProvider(),
  new ImdbMovieSearchProvider(),
];

const gameProviders: GameSearchProvider[] = [
  new RawgGameSearchProvider(),
  new SteamGameSearchProvider(),
  new PsnGameSearchProvider(),
  new XboxGameSearchProvider(),
  new SwitchGameSearchProvider(),
];

// 电视剧搜索 Provider
const tvShowProviders: TvShowSearchProvider[] = [
  new TmdbTvShowSearchProvider(),
];

function parseProviders(providers?: string[]): Set<string> {
  if (!providers || providers.length === 0) return new Set();
  const set = new Set<string>();
  for (const value of providers) {
    if (!value) continue;
    for (const part of value.split(',')) {
      const trimmed = part.trim().toLowerCase();
      if (trimmed) set.add(trimmed);
    }
  }
  return set;
}

function selectProviders<T extends { id(): string }>(all: T[], requested: Set<string>): T[] {
  if (requested.size === 0) return all;
  return all.filter((p) => requested.has(p.id().toLowerCase()));
}

async function safeMovieSearch(
  provider: MovieSearchProvider,
  query: string,
  page: number,
): Promise<ProviderSearchResult<ExternalMovieSearchResult>> {
  try {
    return await provider.search(query, page);
  } catch (ex: any) {
    return {
      provider: provider.id(),
      enabled: true,
      message: `搜索失败: ${ex.message ?? ex}`,
      page,
      totalPages: 0,
      totalResults: 0,
      results: [],
    };
  }
}

async function safeGameSearch(
  provider: GameSearchProvider,
  query: string,
  page: number,
): Promise<ProviderSearchResult<ExternalGameSearchResult>> {
  try {
    return await provider.search(query, page);
  } catch (ex: any) {
    return {
      provider: provider.id(),
      enabled: true,
      message: `搜索失败: ${ex.message ?? ex}`,
      page,
      totalPages: 0,
      totalResults: 0,
      results: [],
    };
  }
}

async function safeTvShowSearch(
  provider: TvShowSearchProvider,
  query: string,
  page: number,
): Promise<ProviderSearchResult<ExternalTvShowSearchResult>> {
  try {
    return await provider.search(query, page);
  } catch (ex: any) {
    return {
      provider: provider.id(),
      enabled: true,
      message: `搜索失败: ${ex.message ?? ex}`,
      page,
      totalPages: 0,
      totalResults: 0,
      results: [],
    };
  }
}

export async function searchMovies(
  query: string,
  page: number,
  providers?: string[],
): Promise<ExternalSearchResponse<ExternalMovieSearchResult>> {
  const requested = parseProviders(providers);
  const selected = selectProviders(movieProviders, requested);
  const providerResults = await Promise.all(
    selected.map((p) => safeMovieSearch(p, query, page)),
  );
  return { query, page, providers: providerResults };
}

export async function searchGames(
  query: string,
  page: number,
  providers?: string[],
): Promise<ExternalSearchResponse<ExternalGameSearchResult>> {
  const requested = parseProviders(providers);
  const selected = selectProviders(gameProviders, requested);
  const providerResults = await Promise.all(
    selected.map((p) => safeGameSearch(p, query, page)),
  );
  return { query, page, providers: providerResults };
}

export async function searchTvShows(
  query: string,
  page: number,
  providers?: string[],
): Promise<ExternalSearchResponse<ExternalTvShowSearchResult>> {
  const requested = parseProviders(providers);
  const selected = selectProviders(tvShowProviders, requested);
  const providerResults = await Promise.all(
    selected.map((p) => safeTvShowSearch(p, query, page)),
  );
  return { query, page, providers: providerResults };
}